src/moments.jl

```diff
@@ -233,7 +236,8 @@ end
 
 function _moment2(v::RealArray, wv::AbstractWeights, m::Real; corrected=false)
     n = length(v)
-    s = 0.0
+    T = promote_type(eltype(v), eltype(wv), typeof(m))
+    s = zero(T)
```

Comment:
To ensure type stability, how about following exactly what the loop does and using `T = typeof(zero(eltype(v) - zero(m))^2 * zero(eltype(w)))`? That's what I did in my first attempt to moving these functions to Statistics: https://github.com/JuliaLang/julia/pull/31395/files#diff-6b82a0c4e60cbe68048e35375987a86688afb4aea60dda8f2d2d4938dc845776R375